docs(route53-patterns): add detail to  README and doc strings (#9194)
aws/aws-cdk: packages/@aws-cdk/aws-route53-patterns/README.md

---
<!--END STABILITY BANNER-->

This library provides higher-level Amazon Route 53 constructs which follow common
architectural patterns.

## HTTPS Redirect

If you want to speed up delivery of your web content, you can use Amazon CloudFront,
the AWS content delivery network (CDN). CloudFront can deliver your entire website
including dynamic, static, streaming, and interactive content—by using a global
network of edge locations. Requests for your content are automatically routed to the
edge location that gives your users the lowest latency.

This construct allows creating a redirect from domainA to domainB using Amazon
CloudFront and Amazon S3. You can specify multiple domains to be redirected.
[Learn more](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-to-cloudfront-distribution.html) about routing traffic to a CloudFront web distribution.

The `HttpsRedirect` constructs creates:

* Amazon CloudFront distribution - makes website available from data centres
  around the world

> * Amazon S3 bucket - used to store content that CloudFront will use as the origin

Reviewer: This construct creates a bucket with a redirect configuration (`websiteRedirect`) that will remain empty. It is only used for the redirection.

(taht's why it uses `RemovalPolicy.DESTROY`)

* Amazon Route 53 Alias record - routes traffic to the CloudFront distribution
* AWS Certificate Manager certificate - X.509 PEM format certificate used by
  CloudFront for your domain

> ⚠️ If you want to require HTTPS between viewers and CloudFront, you must change the AWS Region to US East (N. Virginia) in the AWS Certificate Manager console.

Reviewer 2: > ...in the console.

:)

Maybe worth calling out instead that the certificate will be created in the US East (N. Virginia) region. But given you're using the cross-region DnsValidatedCertificate, the stack/construct as a whole can be anywhere.

> ⚠️ If you want to require HTTPS between viewers and CloudFront, you must change the AWS Region to US East (N. Virginia) in the AWS Certificate Manager console.

Reviewer: *If you want to use an existing certificate it must be created in US East (N. Virginia)...*? We always want HTTPS with this construct.

Author: fair point, too much copy pasta from the AWS docs :)

The following example creates an HTTPS redirect from `foo.example.com` to `bar.example.com`

> It does not provide an ACM certificate, so one will be created in `us-east-1` by the CDK.

Reviewer 2: ```suggestion
As an existing certificate is not provided, one will be created in `us-east-1` by the CDK.
```

This construct allows creating a redirect from domainA to domainB using Amazon
CloudFront and Amazon S3. You can specify multiple domains to be redirected.
[Learn more](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-to-cloudfront-distribution.html) about routing traffic to a CloudFront web distribution.

The `HttpsRedirect` constructs creates:

* Amazon CloudFront distribution - makes website available from data centres
  around the world
* Amazon S3 bucket - empty bucket used for website hosting redirect (`websiteRedirect`) capabilities.
* Amazon Route 53 Alias record - routes traffic to the CloudFront distribution
* AWS Certificate Manager certificate - SSL/TLS certificate used by
  CloudFront for your domain

⚠️ The stack/construct can be used in any region for configuring an HTTPS redirect.
The certificate created in Amazon Certificate Manager (ACM) will be in US East (N. Virginia)
region. If you use an existing certificate, the AWS region of the certificate
must be in US East (N. Virginia).

You cannot migrate an existing certificate in ACM from one AWS Region to another.
To associate an ACM Certificate with a CloudFront distribution, you must create

> a certificate in the US East (N. Virginia)region. [Read more](https://aws.amazon.com/premiumsupport/knowledge-center/migrate-ssl-cert-us-east/) about how to create a certificate.

Reviewer 2: Nit - I'd just remove this paragraph entirely, but up to you.

  });
  ```

The removal policy on the S3 bucket used to store your content is set to `RemovalPolicy.DESTROY`

> When your stack is deleted, the bucket will be physically destroyed.

Reviewer 2: Will everything else as well? Any particular reason to call out the bucket here (again, given there isn't any content in it)?

Author: I guess it isn't needed to explicitly call it out. I'll remove it

Author: the reasoning for including a blurb is because the default for an S3 bucket is to retain on removal. The motivation was to try and be a little transparent about the opinionated setting being used in this pattern.

Reviewer 2: Yeah, I suppose. Counter-argument -- the bucket will have no content, the user gets no reference to it (to add content), and isn't really being used as a bucket so much as a "website redirect mechanism".
